### ISSUE TITLE
Why can't \hfill be used to left- or right-justify displaystyle?

### ISSUE THREAD

**Issue by jason-s** [mathjax/MathJax#1700]
Why doesn't \hfill work to provide left- or right-justified display style?

For example:

    $$A=\pi r^2 \hfill $$

    $$\hfill A=\pi r^2 $$

whereas it does work in certain MathJax environments e.g. this test case from #903

       \[
         \begin{array}{cc}
           aaaaaa&b\hfill\\
           \hfill b&aaaaa\\
         \end{array}
       \]

**Comment by jason-s** [comment]
see also #1075

**Comment by pkra** [comment]
Renamed and marked as feature request.

**Comment by dpvc** [comment]
Note that `$$A=\pi r^2 \hfill $$` does not produce a left-justified equation in LaTeX, either, so I'm not sure I would want to give `\hfill` that non-standard interpretation.  It would be better to provide a separate command for left and right-justified expressions.

One solution at the moment would be to use the `multilane` environment with `\shoveleft` or `\shoveright`.  E.g.,

```
\begin{multline}
\shoveleft A = \pi r^2
\end{multilane}
```

however, the default for multline in MathJax is for it to use 85% of the width.  If you have control over the page's configuration, you can change that using

```
MathJax.Hub.Config({
  TeX: {
  }
});

**Comment by dpvc** [comment]
(Sorry, accidentally hit "comment" too soon via keyboard equivalents.  I've updated the comment above.)

**Comment by tavin** [comment]
How could this be implemented? I just want to display certain symbols on the right in markdown.

**Comment by dpvc** [comment]
@tavin, `\hfill` is not the way to do that even in true LaTeX.

Markdown does not provide a means of right-justification, but you can insert HTML into markdown, so could do

``` html
<div style="text-align: right">
$ ... (your symbol here) ... $
</div>
```

Otherwise, the `multline` approach with `\shoveright` is the best solution I can suggest.

**Comment by tavin** [comment]
Hi @dpvc, thanks for the information. I suppose it may not be considered a best practice to use `\hfill` in this way, in true LateX. But would it be harmful to add this feature to MathJax? I looked through the sources and it seems there is a lot of support for `\hfill` -- at least for arrays. So maybe it's not so difficult to add this use case? I wouldn't know the right or wrong way to go about it in MathJax so that's what I'm trying to ask. Cheers.

**Comment by dpvc** [comment]
@tavin, although MathJax does have some features that are different from LaTeX (and even some that are intentionally different), most of those come either from practical considerations, or from decisions early on in MathJax's design before we had as much information non the use cases that MathJax is being put to.  At this point we are trying to limit the differences between MathJax and LaTeX rather than introduce new ones.  One of the reasons for this is that many people move material between actual LaTeX and MathJax and vice versa, either boy hand, or through automated processes like `tex4ht`, `panda`, etc., and inconsistencies cause trouble with that.  Not everyone will be careful about separating the two environments.

On the other hand, TeX-based control of the centering *is* something we would like to allow, but `\hfill` is probably not how it is going to be handled in MathJax.

**Comment by tavin** [comment]
@dpvc sounds like we are in complete agreement. Shouldn't we eliminate one difference between MathJax and LaTeX by implementing `\hfill`? What users may choose to do with it is beside the point, actually -- I shouldn't have brought up that aspect in the first place.

**Comment by dpvc** [comment]
@tavin, there are lots of piecers of LaTeX that MathJax doesn't implement, and it is not practical for MathJax to do them all.  MathJax uses MathML as its internal format, and MathML doesn't have an idea of stretchable glue like TeX's `\hfill` and similar macros.  That makes it difficult to provide a general implementation for `\hfill`.  As you note, MathJax already handles `\hfill` in arrays in its most common usages.  While it would be possible to extend that to support for some other use-cases, we have limited resources, and have to apply them where they are most needed, especially by our sponsors.  There are quite a backlog of features that are still needed that would likely come before that (the current `\hfill` implementation came from the needs of one of our sponsors).

On the other hand, if a community member wished to make such an implementation, we would certainly consider a PR that provides it.

**Comment by tavin** [comment]
@dpvc yes that is perfectly reasonable. I was only hoping for a clue about how to get started on such an implementation. From your last comment I gather it would not be easy after all.
